modules/OFConnectionManager2/module/src/bundle.c

```diff
@@ -262,23 +331,69 @@ parse_message(uint8_t *data, of_object_storage_t *storage)
     return of_object_new_from_message_preallocated(storage, data, len);
 }
 
+
+static void
+free_subbundle(subbundle_t *subbundle)
+{
+    int i;
+    for (i = 0; i < subbundle->count; i++) {
+        aim_free(subbundle->msgs[i]);
+    }
+    aim_free(subbundle->msgs);
+}
+
 static void
 free_bundle(bundle_t *bundle)
 {
     int i;
-    for (i = 0; i < bundle->count; i++) {
-        aim_free(bundle->msgs[i]);
+    for (i = 0; i < bundle->subbundle_count; i++) {
+        free_subbundle(&bundle->subbundles[i]);
     }
-
-    aim_free(bundle->msgs);
+    aim_free(bundle->subbundles);
 
     bundle->id = BUNDLE_ID_INVALID;
-    bundle->msgs = NULL;
+    bundle->subbundles = NULL;
+    bundle->subbundle_count = 0;
     bundle->count = 0;
-    bundle->allocated = 0;
     bundle->bytes = 0;
     bundle->flags = 0;
+}
+
+/* see description in public header file */
+void
+indigo_cxn_subbundle_set(uint32_t num_subbundles,
+                         indigo_cxn_subbundle_designator_t designator,
+                         indigo_cxn_bundle_comparator_t comparators[])
+{
+    /* one more subbundle for barriers */
+    num_subbundles_per_bundle = num_subbundles + 1;
+    if (num_subbundles_per_bundle >
+        OFCONNECTIONMANAGER_CONFIG_MAX_SUBBUNDLES+1) {
+        num_subbundles_per_bundle =
+            OFCONNECTIONMANAGER_CONFIG_MAX_SUBBUNDLES+1;
+    }
+    subbundle_designator = designator;
+    subbundle_comparators = comparators;
```

Comment:
Please make sure user supplied num_subbundles_per_bundle number of subbundle_comparators, if it is not NULL.